x/derivatives/keeper/perpetual_futures.go

```diff
@@ -77,54 +77,28 @@ func (k Keeper) ClosePerpetualFuturesPosition(ctx sdk.Context, position types.Po
 	feeAmount := positionInstance.Size_.Mul(commissionRate)
 	tradeAmount := positionInstance.Size_.Sub(feeAmount)
 
-	// todo: fixme
-	openedRate := position.OpenedBaseRate
 	closedRate, err := k.GetPairRate(ctx, position.Market)
 	if err != nil {
 		return err
 	}
 
-	k.bankKeeper.SendCoinsFromAccountToModule(ctx, position.Address.AccAddress(), types.ModuleName, sdk.Coins{sdk.NewCoin(position.Market.BaseDenom, feeAmount.RoundInt())}) // TODO: this is wrong.
-
-	principal := positionInstance.CalculatePrincipal()
-	amountToUser := sdk.Dec{}
-
-	switch positionInstance.PositionType {
-	case types.PositionType_LONG:
-		k.SubPerpetualFuturesNetPositionOfMarket(ctx, position.Market, tradeAmount)
-
-		if closedRate.GTE(openedRate) {
-			profit := closedRate.Mul(sdk.NewDec(int64(positionInstance.Leverage))).Sub(positionInstance.Size_)
-			profitAmount := profit.Quo(*closedRate)
-
-			amountToUser = principal.Add(profitAmount)
-		} else {
-			loss := positionInstance.Size_.Sub(closedRate.Mul(sdk.NewDec(int64(positionInstance.Leverage))))
-			lossAmount := loss.Quo(*closedRate)
-
-			amountToUser = principal.Sub(lossAmount)
-		}
-		break
-	case types.PositionType_SHORT:
-		k.AddPerpetualFuturesNetPositionOfMarket(ctx, position.Market, tradeAmount)
-
-		if closedRate.LTE(openedRate) {
-			profit := positionInstance.Size_.Sub(closedRate.Mul(sdk.NewDec(int64(positionInstance.Leverage))))
-			profitAmount := profit.Quo(*closedRate)
-
-			amountToUser = principal.Add(profitAmount)
-		} else {
-			loss := closedRate.Mul(sdk.NewDec(int64(positionInstance.Leverage))).Sub(positionInstance.Size_)
-			lossAmount := loss.Quo(*closedRate)
+	err = k.bankKeeper.SendCoinsFromAccountToModule(ctx, position.Address.AccAddress(), types.ModuleName, sdk.Coins{sdk.NewCoin(position.Market.BaseDenom, feeAmount.RoundInt())}) // TODO: this is wrong.
+	if err != nil {
+		return err
+	}
 
-			amountToUser = principal.Sub(lossAmount)
-		}
-		break
-	case types.PositionType_POSITION_UNKNOWN:
-		return fmt.Errorf("unknown position type")
+	futuresPosition := types.NewPerpetualFuturesPosition(position, positionInstance)
+	returningAmount := futuresPosition.CalcReturningAmountAtClose(*closedRate)
+	// If loss is over the margin, it means liquidity provider takes the loss.
+	if returningAmount.IsNegative() {
+		returningAmount = sdk.ZeroInt()
+		// emit event to tell how much loss is taken by liquidity provider.
```

Comment:
plz move this code to CalcReturningAmountAtClose.
Business logic should be confined within types.

Reply:
then, what's the best way that we treat the amount of the loss to the liquidity provider?
add a returning value containing possible loss to liquidity provider?

Reply:
> add a returning value containing possible loss to liquidity provider?

please do it

Reply:
done.
plz check